### ISSUE TITLE
Make back-end of Operator: RayTransform compatible with Tensorflow-gpu

### ISSUE THREAD

**Issue by ginsanity15** [odlgroup/odl#1482]
I was trying to perform tomographic reconstruction with ODL (version 1.0.0) and Tensorflow-gpu (version 1.13.1) but I have encountered a dilemma as follow. 

In order to utilize GPU for the computation of forward and backward projection in ODL, it is mandatory to install "Astra-toolbox". However, installing "Astra-toolbox" will downgrade package "numpy" to version 1.12. If we import tensorflow to python in this situation, the following error message will occur:

> ModuleNotFoundError: No module named 'numpy.core._multiarray_umath'
> ImportError: numpy.core.multiarray failed to import
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "<frozen importlib._bootstrap>", line 968, in _find_and_load
> SystemError: <class '_frozen_importlib._ModuleLockManager'> returned a result with an error set
> ImportError: numpy.core._multiarray_umath failed to import
> ImportError: numpy.core.umath failed to import

I have tried to upgrade "numpy" to version 1.16 after installing "astra-toolbox", but then ODL will complain there is no available back-end for Ray-Transform. 

Have you met this problem before? If so, could you please tell me how could I solve it? Thanks for your attention, I am looking forward to hear back from you.

**Comment by kohr-h** [comment]
Thanks for reporting this issue @ginsanity15. This is outside our control unfortunately. What you can try is to install ASTRA from @wjp's conda channel: `conda install astra-toolbox -c wjpalenstijn`. It has development versions that depend on newer versions of NumPy.

**Comment by wjp** [comment]
I recommend using the `astra-toolbox/label/dev` channel now instead of my personal channel.

**Comment by kohr-h** [comment]
Even better. Thanks @wjp!

**Comment by ginsanity15** [comment]
Thanks for your replies but I have to say I opened this issue here also because it is impossible to import `odl.contrib.tensorflow` now. 

Anyway, I tried to install the development version of astra toolbox but that doesn't solve the problem. I guess the best thing I can do is to downgrade tensorflow. There should be a way to solve it as I have read two papers written by @adler-j that create tensor layer with RayTransform in ODL. 

Thanks again for your reply, I will keep you updated if I got a solution.

**Comment by ginsanity15** [comment]
Okay, I think I may have found a solution. Under the setup listed below, it is possible to make astra toolbox and tensorflor-gpu compatible with each other. More importantly, we could import `odl.contrib.tensorflow` now, hooray!

CUDA toolkit: 9.0
cuDNN: 7.0.4
Python: 3.6.7
Numpy: 1.15.4
TensorFlow-gpu: 1.12.0
astra toolbox: development version
ODL: 1.0.0

Hope this could be helpful to someone else in the future and thanks again for answering my questions.

**Comment by kohr-h** [comment]
Great you figured it out @ginsanity15! Apart from the versions of ASTRA and NumPy, what else did you change? The CUDA version maybe?

**Comment by kohr-h** [comment]
Ah, I see that the TF version is lower. Could it be that `odl.contrib.tensorflow` does not work with TF 1.13 and higher?

**Comment by ginsanity15** [comment]
Yep, TF is now 1.12, and in this version, we need CUDA toolkit 9.0 instead of 10.0. The version of cuDNN should be 7.0.4 instead of 7.4. 

As for `odl.contrib.tensorflow`, it actually works with TF 1.13 (haven't tested higher version). I suspect the problem is: TF 1.13 requires numpy version to be 1.16 but astra-toolbox will downgrade numpy to 1.15.4. The lower version of numpy makes TF impossible to run and messes up the ODL tensorflow part as well.

I haven't read source code for `odl.contrib.tensorflow` yet so everything above is just a guess...

**Comment by kohr-h** [comment]
I just tried to reproduce your issue, but I didn't encounter any problem whatsoever. I ran an example from `odl.contrib.tensorflow`, and it worked.

Regarding NumPy dependency, `tensorflow-gpu` 1.13.1 requires `numpy>=1.13.3` so there's no conflict with the versions that ASTRA or ODL require.

If you first install `astra-toolbox` using `conda install astra-toolbox -c astra-toolbox/label/dev` and then `odl` and `tensorflow-gpu`, it should just work.
